File: apps/mobile/ios/Sources/FlagshipUI/ViewModels/CompanionDockViewModel.swift
import Foundation
import Observation
import Flagship
import FlagshipAPI
import FlagshipCore

/// Backs the phone side of desktop-initiated companion docking.
@MainActor
@Observable
public final class CompanionDockViewModel {
    public private(set) var state: LoadingState<CompanionListResponse> = .idle
    public private(set) var stagedApproval: CompanionDockApprovalLink.Payload?
    public private(set) var approvalPending = false
    public private(set) var approvalError: String?
    public private(set) var approvalComplete = false
    public private(set) var revokePending: Set<String> = []

    private let client: any ScreensClient
    private let expectedServerDomain: String?
    private let authenticate: @MainActor @Sendable (String) async throws -> Void

    public init(
        client: any ScreensClient,
        expectedServerDomain: String? = nil,
        authenticate: @escaping @MainActor @Sendable (String) async throws -> Void = { reason in
            try await BiometricGate().evaluate(reason: reason)
        }
    ) {
        self.client = client
        self.expectedServerDomain = expectedServerDomain?.lowercased()
        self.authenticate = authenticate
    }

    public func load() async {
        state = .loading
        do {
            state = .loaded(try await client.companionList())
        } catch {
            state = .failed(error.localizedDescription)
        }
    }

    @discardableResult
    public func stageApproval(link raw: String) -> Bool {
        approvalError = nil
        approvalComplete = false
        guard let parsed = CompanionDockApprovalLink.parse(raw) else {
            stagedApproval = nil
            approvalError = "This isn't a valid Flagship docking link."
            return false
        }
        if let expectedServerDomain, parsed.serverDomain != expectedServerDomain {
            stagedApproval = nil
            approvalError = "Switch to \(parsed.serverDomain) in Flagship, then scan this code again."
            return false
        }
        stagedApproval = parsed
        return true
    }

    public func clearApproval() {
        stagedApproval = nil
        approvalError = nil
        approvalComplete = false
    }

    public func approve() async {
        guard let stagedApproval, !approvalPending else { return }
        approvalPending = true
        approvalError = nil
        defer { approvalPending = false }
        do {
            try await authenticate("Dock this browser to your Flagship account")
            _ = try await client.companionApproveDock(
                CompanionDockApproveRequest(
                    requestId: stagedApproval.requestId,
                    approvalSecret: stagedApproval.approvalSecret
                )
            )
            approvalComplete = true
            self.stagedApproval = nil
            await load()
        } catch {
            approvalError = error.localizedDescription
        }
    }

    public func revoke(tokenPrefix: String) async {
        revokePending.insert(tokenPrefix)
        defer { revokePending.remove(tokenPrefix) }
        do {
            _ = try await client.companionRevoke(
                CompanionRevokeRequest(tokenPrefix: tokenPrefix)
            )
            if case .loaded(let list) = state {
                state = .loaded(
                    CompanionListResponse(
                        companions: list.companions.filter { $0.tokenPrefix != tokenPrefix }
                    )
                )
            }
        } catch {
            state = .failed(error.localizedDescription)
        }
    }
}

File: apps/mobile/android/app/src/main/java/com/flagshipserver/app/viewmodels/CompanionDockViewModel.kt
// Drives the phone side of desktop-initiated docking: scan/paste, explicit
// biometric approval, active-companion listing, and per-row revocation.

package com.flagshipserver.app.viewmodels

import com.flagshipserver.app.api.CompanionListResponse
import com.flagshipserver.app.api.CompanionDockApproveRequest
import com.flagshipserver.app.api.CompanionRevokeRequest
import com.flagshipserver.app.api.ScreensClient
import com.flagshipserver.app.api.ScreensError
import com.flagshipserver.app.core.NetworkErrorHumanizer
import com.flagshipserver.app.core.CompanionDockApprovalLink
import com.flagshipserver.app.core.CompanionDockApprovalPayload
import com.flagshipserver.app.keystore.BiometricAuthority
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

class CompanionDockViewModel(
    private val client: ScreensClient,
    private val scope: CoroutineScope = CoroutineScope(SupervisorJob() + Dispatchers.IO),
    private val expectedServerDomain: String? = null,
    private val authenticate: suspend (String) -> Unit = { reason ->
        BiometricAuthority.current()?.let { authority ->
            // Dock approval is a fresh, explicit ceremony even when the rest
            // of the unlocked app session has a warm biometric latch.
            authority.invalidate()
            authority.ensureFresh("Dock this browser", reason)
        }
    },
) {
    private val _state = MutableStateFlow<LoadingState<CompanionListResponse>>(LoadingState.Idle)
    val state: StateFlow<LoadingState<CompanionListResponse>> = _state.asStateFlow()

    private val _stagedApproval = MutableStateFlow<CompanionDockApprovalPayload?>(null)
    val stagedApproval: StateFlow<CompanionDockApprovalPayload?> = _stagedApproval.asStateFlow()

    private val _approvalPending = MutableStateFlow(false)
    val approvalPending: StateFlow<Boolean> = _approvalPending.asStateFlow()

    private val _approvalError = MutableStateFlow<String?>(null)
    val approvalError: StateFlow<String?> = _approvalError.asStateFlow()

    private val _approvalComplete = MutableStateFlow(false)
    val approvalComplete: StateFlow<Boolean> = _approvalComplete.asStateFlow()

    private val _revokePending = MutableStateFlow<Set<String>>(emptySet())
    val revokePending: StateFlow<Set<String>> = _revokePending.asStateFlow()

    fun load(): Job = scope.launch {
        _state.value = LoadingState.Loading
        runCatching { client.companionList() }
            .fold(
                onSuccess = { _state.value = LoadingState.Loaded(it) },
                onFailure = { t -> _state.value = LoadingState.Failed(failureMessage(t)) },
            )
    }

    fun stageApproval(raw: String): Boolean {
        _approvalError.value = null
        _approvalComplete.value = false
        val parsed = CompanionDockApprovalLink.parse(raw)
        if (parsed == null) {
            _stagedApproval.value = null
            _approvalError.value = "This isn't a valid Flagship docking link."
            return false
        }
        if (expectedServerDomain != null && parsed.serverDomain != expectedServerDomain.lowercase()) {
            _stagedApproval.value = null
            _approvalError.value = "Switch to ${parsed.serverDomain} in Flagship, then scan this code again."
            return false
        }
        _stagedApproval.value = parsed
        return true
    }

    fun clearApproval() {
        _stagedApproval.value = null
        _approvalError.value = null
        _approvalComplete.value = false
    }

    fun approve(): Job = scope.launch {
        val approval = _stagedApproval.value ?: return@launch
        if (_approvalPending.value) return@launch
        _approvalPending.value = true
        _approvalError.value = null
        try {
            authenticate("Approve a four-hour keyless companion session")
            client.companionApproveDock(
                CompanionDockApproveRequest(
                    requestId = approval.requestId,
                    approvalSecret = approval.approvalSecret,
                ),
            )
            _approvalComplete.value = true
            _stagedApproval.value = null
            runCatching { client.companionList() }
                .onSuccess { _state.value = LoadingState.Loaded(it) }
        } catch (t: Throwable) {
            _approvalError.value = failureMessage(t)
        } finally {
            _approvalPending.value = false
        }
    }

    fun revoke(tokenPrefix: String): Job = scope.launch {
        _revokePending.value = _revokePending.value + tokenPrefix
        try {
            runCatching { client.companionRevoke(CompanionRevokeRequest(tokenPrefix = tokenPrefix)) }
                .fold(
                    onSuccess = {
                        val current = _state.value
                        if (current is LoadingState.Loaded) {
                            val remaining = current.value.companions.filterNot { it.tokenPrefix == tokenPrefix }
                            _state.value = LoadingState.Loaded(CompanionListResponse(companions = remaining))
                        }
                    },
                    onFailure = { t -> _state.value = LoadingState.Failed(failureMessage(t)) },
                )
        } finally {
            _revokePending.value = _revokePending.value - tokenPrefix
        }
    }

    private fun failureMessage(t: Throwable): String = when (t) {
        is ScreensError.Http -> NetworkErrorHumanizer.humanize(t)
        else -> t.message ?: "couldn't load companions"
    }
}
